### PR TITLE
Linter - add tests

### PR DESCRIPTION
**Pull request by keradus**

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Tests failing?

File: Symfony/CS/Tests/Linter/LinterTest.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+ * @internal
+ */
+final class LinterTest extends \PHPUnit_Framework_TestCase
+{
+    /**
+     * @runInSeparateProcess
+     */
+    public function testPrepareCommand()
+    {
+        $fixEscape = ('\'' === substr(ProcessUtils::escapeArgument('x'), 0, 1)
```
Comment (GrahamCampbell) [inline Symfony/CS/Tests/Linter/LinterTest.php]: this ternary statement is very hard to read

File: Symfony/CS/Tests/Linter/LinterTest.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    /**
+     * @expectedException Symfony\CS\Linter\LintingException
+     */
+    public function testLintSourceWithBadCode()
+    {
+        $linter = new Linter();
+        $linter->lintSource('<?php echo echo;');
+    }
+
+    private function invokeMethod(&$object, $methodName, array $parameters = array())
```
Comment (GrahamCampbell) [inline Symfony/CS/Tests/Linter/LinterTest.php]: don't need a `&` surely if that param is an actual object?

File: Symfony/CS/Linter/Linter.php [hunk 3]
```diff
@@ -135,4 +134,16 @@ private function createProcessForSource($source)
 
         return $process;
     }
+
+    /**
+     * @param string $path
+     *
+     * @return string
+     */
+    private function prepareCommand($path)
```
Comment (GrahamCampbell) [inline Symfony/CS/Linter/Linter.php]: good idea

**Comment by keradus** [comment]
Tests fails, yes.
Fails on PHP* are present due to bug on your branch that I want to prove by this PR. It is the point...
Fail on HHVM is... WTF? dont know what's going on with HHVM and phpunit...

File: Symfony/CS/Tests/Linter/LinterTest.php (hunk 2 above)
Comment (keradus) [inline Symfony/CS/Tests/Linter/LinterTest.php]: yeah!

**Comment by keradus** [comment]
To make that tests I had to change few other things, extracted here:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1405/files

File: Symfony/CS/Tests/Linter/LinterTest.php (hunk 1 above)
Comment (keradus) [inline Symfony/CS/Tests/Linter/LinterTest.php]:
```php
$a = ($b
    ? $c
    : $d
);
```

clean for me ;)

**Comment by keradus** [comment]
updated after your rebase!
great to see this happen ;)

**Comment by GrahamCampbell** [comment]
> PHPUnit_Framework_Exception: Fatal error: Class undefined: PHP_Token_HASHBANG in phar://phpunit-4.5.0.phar/php-token-stream/Token/Stream.php on line 185

That can be fixed by updating phpunit. I've seen that error before.

File: Symfony/CS/Tests/Linter/LinterTest.php [hunk 4]
```diff
@@ -21,6 +22,43 @@
...
+        $fixEscape = ('\'' === substr(ProcessUtils::escapeArgument('x'), 0, 1)
+            ? function ($str) {
+                return str_replace('"', '\'', $str);
+            }
+            : function ($str) {
+                return $str;
+            }
+        );
+
+        // Case 1: standard call
```
Comment (GrahamCampbell) [inline Symfony/CS/Tests/Linter/LinterTest.php]: this test is wrong on hhvm

File: Symfony/CS/Tests/Linter/LinterTest.php [hunk 5]
```diff
@@ -21,6 +22,43 @@
...
+        $this->assertSame(
+            $fixEscape('"php" -l "foo.php"'),
+            $this->invokeMethod(new Linter('php'), 'prepareCommand', array('foo.php'))
+        );
+
+        // Case 2: call with space in dir
```
Comment (GrahamCampbell) [inline Symfony/CS/Tests/Linter/LinterTest.php]: this test is wrong on hhvm

**Comment by keradus** [comment]
Damn it :/
So that error should be fixed by Travis, we may do a workaround and download our own version :(

File: Symfony/CS/Tests/Linter/LinterTest.php (hunk 5 above)
Comment (keradus) [inline Symfony/CS/Tests/Linter/LinterTest.php]: good catch

Comment (GrahamCampbell) [inline Symfony/CS/Tests/Linter/LinterTest.php]: Maybe better to just have 3 different tests that require different runtimes, like we already do for some other stuff

Comment (keradus) [inline Symfony/CS/Tests/Linter/LinterTest.php]: or not :D
do you know why HHVM on Travis doesn't use `--php` arg?

Comment (GrahamCampbell) [inline Symfony/CS/Tests/Linter/LinterTest.php]: I think that argument is new in 3.6.0 or something.

**Comment by keradus** [comment]
I think this PR is ready. The only bug reported by PHPUnit is a bug that this PR was supposed to detect and prove

**Comment by keradus** [comment]
ping @GrahamCampbell

**Pull request merged**